File: src/main/java/com/joecollins/bindings/IndexedBinding.java
package com.joecollins.bindings;

import java.util.List;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.IntFunction;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public interface IndexedBinding<T> {

  int size();

  T getValue(int index);

  default void bind(BiConsumer<Integer, T> onUpdate) {
    IntStream.range(0, size()).forEach(idx -> onUpdate.accept(idx, getValue(idx)));
  }

  default void unbind() {}

  default <R> IndexedBinding<R> map(Function<T, R> func) {
    IndexedBinding<T> me = this;
    return new IndexedBinding<R>() {
      @Override
      public int size() {
        return me.size();
      }

      @Override
      public R getValue(int index) {
        return func.apply(me.getValue(index));
      }

      @Override
      public void bind(BiConsumer<Integer, R> onUpdate) {
        me.bind((idx, val) -> onUpdate.accept(idx, func.apply(val)));
      }

      @Override
      public void unbind() {
        me.unbind();
      }
    };
  }

  static <T> IndexedBinding<T> emptyBinding() {
    return new IndexedBinding<>() {
      @Override
      public int size() {
        return 0;
      }

      @Override
      public T getValue(int index) {
        return null;
      }
    };
  }

  static <T> IndexedBinding<T> singletonBinding(T item) {
    return new IndexedBinding<>() {
      @Override
      public int size() {
        return 1;
      }

      @Override
      public T getValue(int index) {
        return item;
      }
    };
  }

  static <T> IndexedBinding<T> functionBinding(
      int startInclusive, int endExclusive, IntFunction<T> itemFunc) {
    return new IndexedBinding<>() {
      @Override
      public int size() {
        return endExclusive - startInclusive;
      }

      @Override
      public T getValue(int index) {
        return itemFunc.apply(startInclusive + index);
      }
    };
  }

  static <T> IndexedBinding<T> listBinding(T... items) {
    return new IndexedBinding<>() {
      @Override
      public int size() {
        return items.length;
      }

      @Override
      public T getValue(int index) {
        return items[index];
      }
    };
  }

  static <T> IndexedBinding<T> listBinding(List<T> items) {
    return new IndexedBinding<>() {
      @Override
      public int size() {
        return items.size();
      }

      @Override
      public T getValue(int index) {
        return items.get(index);
      }
    };
  }

  static <E extends Enum<E>, T extends Bindable<E>, U> IndexedBinding<U> propertyBinding(
      T item, Function<T, List<? extends U>> func, E... properties) {
    if (item == null) {
      return emptyBinding();
    }
    return new IndexedBinding<>() {
      private Consumer<T> consumer = null;

      @Override
      public U getValue(int idx) {
        if (idx < 0 || idx >= size()) {
          return null;
        }
        return func.apply(item).get(idx);
      }

      @Override
      public int size() {
        return func.apply(item).size();
      }

      @Override
      public void bind(BiConsumer<Integer, U> onUpdate) {
        if (consumer != null) {
          throw new IllegalStateException("Binding is already used");
        }
        consumer =
            val -> {
              List<? extends U> vals = func.apply(val);
              for (int i = 0; i < vals.size(); i++) {
                onUpdate.accept(i, vals.get(i));
              }
            };
        item.addBinding(consumer, properties);
        consumer.accept(item);
      }

      @Override
      public void unbind() {
        if (consumer == null) {
          throw new IllegalStateException("Binding is not currently used");
        }
        item.removeBinding(consumer, properties);
        consumer = null;
      }
    };
  }

  static <T> IndexedBinding<T> propertyBinding(BindableList<T> list) {
    return propertyBinding(list, Function.identity());
  }

  static <T, U> IndexedBinding<U> propertyBinding(BindableList<T> list, Function<T, U> func) {
    if (list == null) {
      return emptyBinding();
    }
    return new IndexedBinding<>() {
      private BiConsumer<Integer, T> consumer = null;

      @Override
      public U getValue(int idx) {
        if (idx < 0 || idx >= size()) {
          return null;
        }
        return func.apply(list.get(idx));
      }

      @Override
      public int size() {
        return list.size();
      }

      @Override
      public void bind(BiConsumer<Integer, U> onUpdate) {
        if (consumer != null) {
          throw new IllegalStateException("Binding is already used");
        }
        consumer = (idx, val) -> onUpdate.accept(idx, func.apply(val));
        list.addItemBinding(consumer);
        IntStream.range(0, size()).forEach(idx -> onUpdate.accept(idx, getValue(idx)));
      }

      @Override
      public void unbind() {
        if (consumer == null) {
          throw new IllegalStateException("Binding is not currently used");
        }
        list.removeItemBinding(consumer);
        consumer = null;
      }
    };
  }

  static <T extends Bindable<E>, U, E extends Enum<E>> IndexedBinding<U> propertyBinding(
      NestedBindableList<T, E> list, Function<T, U> func, E... properties) {
    if (list == null) {
      return emptyBinding();
    }
    return new IndexedBinding<>() {
      private BiConsumer<Integer, T> consumer = null;

      @Override
      public U getValue(int idx) {
        if (idx < 0 || idx >= size()) {
          return null;
        }
        return func.apply(list.get(idx));
      }

      @Override
      public int size() {
        return list.size();
      }

      @Override
      public void bind(BiConsumer<Integer, U> onUpdate) {
        if (consumer != null) {
          throw new IllegalStateException("Binding is already used");
        }
        consumer = (idx, val) -> onUpdate.accept(idx, func.apply(val));
        list.addNestedItemBinding(consumer, properties);
        IntStream.range(0, size()).forEach(idx -> onUpdate.accept(idx, getValue(idx)));
      }

      @Override
      public void unbind() {
        if (consumer == null) {
          throw new IllegalStateException("Binding is not currently used");
        }
        list.removeNestedItemBinding(consumer, properties);
        consumer = null;
      }
    };
  }

  static <T, U> IndexedBinding<U> listBinding(List<T> list, Function<T, Binding<U>> bindingFunc) {
    if (list == null) {
      return emptyBinding();
    }
    List<Binding<U>> bindings = list.stream().map(bindingFunc).collect(Collectors.toList());
    return new IndexedBinding<>() {
      @Override
      public int size() {
        return bindings.size();
      }

      @Override
      public U getValue(int index) {
        return bindings.get(index).getValue();
      }

      @Override
      public void bind(BiConsumer<Integer, U> onUpdate) {
        for (int i = 0; i < bindings.size(); i++) {
          int index = i;
          bindings.get(index).bindLegacy(val -> onUpdate.accept(index, val));
        }
      }

      @Override
      public void unbind() {
        bindings.forEach(Binding::unbind);
      }
    };
  }
}

File: src/main/java/com/joecollins/bindings/IndexedBinding.kt
package com.joecollins.bindings

import java.lang.IllegalStateException

interface IndexedBinding<out T> {

    @Deprecated("Convert to Kotlin version")
    @JvmDefault fun bindLegacy(onUpdate: java.util.function.BiConsumer<Int, in T>) = bind { i, v -> onUpdate.accept(i, v) }

    fun bind(onUpdate: (Int, T) -> Unit)

    fun unbind()

    @JvmDefault fun <U> map(func: (T) -> U): IndexedBinding<U> {
        val me = this
        return object : IndexedBinding<U> {
            override fun bind(onUpdate: (Int, U) -> Unit) = me.bind { idx, value -> onUpdate(idx, func(value)) }
            override fun unbind() = me.unbind()
        }
    }

    companion object {
        @JvmStatic fun <T> emptyBinding(): IndexedBinding<T> = object : IndexedBinding<T> {
            override fun bind(onUpdate: (Int, T) -> Unit) {}
            override fun unbind() {}
        }

        @JvmStatic fun <T> singletonBinding(item: T): IndexedBinding<T> = object : IndexedBinding<T> {
            override fun bind(onUpdate: (Int, T) -> Unit) { onUpdate(0, item) }
            override fun unbind() {}
        }

        @JvmStatic fun <T> listBinding(vararg list: T): IndexedBinding<T> = object : IndexedBinding<T> {
            override fun bind(onUpdate: (Int, T) -> Unit) {
                list.indices.forEach { onUpdate(it, list[it]) }
            }
            override fun unbind() {}
        }

        @JvmStatic fun <T> listBinding(list: List<T>): IndexedBinding<T> = object : IndexedBinding<T> {
            override fun bind(onUpdate: (Int, T) -> Unit) {
                list.indices.forEach { onUpdate(it, list[it]) }
            }
            override fun unbind() {}
        }

        @JvmStatic fun <T, U> listBinding(list: List<T>, func: (T) -> Binding<U>): IndexedBinding<U> {
            val bindings = list.map(func)
            return object : IndexedBinding<U> {
                override fun bind(onUpdate: (Int, U) -> Unit) {
                    list.indices.forEach { idx -> bindings[idx].bind { onUpdate(idx, it) } }
                }
                override fun unbind() {
                    bindings.forEach { it.unbind() }
                }
            }
        }

        @JvmStatic fun <E : Enum<E>, T : Bindable<T, E>, U> propertyBinding(item: T, func: (T) -> List<U>, vararg properties: E) = object : IndexedBinding<U> {
            private var consumer: ((T) -> Unit)? = null

            override fun bind(onUpdate: (Int, U) -> Unit) {
                check(consumer == null) { "Binding is already used" }
                val consumer: (T) -> Unit = {
                    val values = func(it)
                    values.indices.forEach { idx -> onUpdate(idx, values[idx]) }
                }
                item.addBinding(consumer, *properties)
                consumer(item)
                this.consumer = consumer
            }
            override fun unbind() {
                val consumer = this.consumer
                if (consumer != null) {
                    item.removeBinding(consumer, *properties)
                    this.consumer = null
                }
            }
        }

        @JvmStatic fun <T> propertyBinding(list: BindableList<T>) = propertyBinding(list) { it }

        @JvmStatic fun <T, U> propertyBinding(list: BindableList<T>, func: (T) -> U) = object : IndexedBinding<U> {
            private var consumer: ((Int, T) -> Unit)? = null

            override fun bind(onUpdate: (Int, U) -> Unit) {
                check(consumer == null) { "Binding is already used" }
                val consumer: (Int, T) -> Unit = { i, t -> onUpdate(i, func(t)) }
                list.addItemBinding(consumer)
                list.indices.forEach { consumer(it, list[it]) }
                this.consumer = consumer
            }
            override fun unbind() {
                val consumer = this.consumer
                if (consumer != null) {
                    list.removeItemBinding(consumer)
                    this.consumer = null
                }
            }
        }

        @JvmStatic fun <E : Enum<E>, T : Bindable<T, E>, U> propertyBinding(list: NestedBindableList<T, E>, func: (T) -> U, vararg properties: E) = object : IndexedBinding<U> {
            private var consumer: ((Int, T) -> Unit)? = null

            override fun bind(onUpdate: (Int, U) -> Unit) {
                check(consumer == null) { throw IllegalStateException("Binding is already used") }
                val consumer: (Int, T) -> Unit = { i, t -> onUpdate(i, func(t)) }
                list.addNestedItemBinding(consumer, *properties)
                list.indices.forEach { consumer(it, list[it]) }
                this.consumer = consumer
            }
            override fun unbind() {
                val consumer = this.consumer
                if (consumer != null) {
                    list.removeNestedItemBinding(consumer, *properties)
                }
                this.consumer = null
            }
        }

        @JvmStatic fun <T> functionBinding(startInclusive: Int, endExclusive: Int, func: (Int) -> T): IndexedBinding<T> = object : IndexedBinding<T> {
            override fun bind(onUpdate: (Int, T) -> Unit) {
                (startInclusive until endExclusive).forEach { onUpdate(it - startInclusive, func(it)) }
            }
            override fun unbind() {}
        }
    }
}
